### PR TITLE
[tflite] Fix for the test interpreter_test.py with Python3

### PR DESCRIPTION
**Pull request by akarmi**
The test interpreter_test.py is failing when it is run using Python3.
There are several issues due to difference between Python2 and Python3:

The string in Python2 is stored as bytes by default, but as unicode in Python3 by default.
In Python3 if the string is in bytes, it has the prefix 'b'.
We convert all strings to be the same.

There is an exception chaining in Python3. As result, the message thrown from the exception is different.

There is a segmentation fault on Python3, when all tests in the file interpreter_test are run. The bug is that the global variable stores the pointer to the function, which goes away from the scope.

The extract from the sample failure log, when running //tensorflow/lite/python:interpreter_test test:

> INFO: From Testing //tensorflow/lite/python:interpreter_test:
> 20:08:53  ==================== Test output for //tensorflow/lite/python:interpreter_test:
> 20:08:53  Running tests under Python 3.6.8: /usr/local/bin/python
> 20:08:53  [ RUN      ] InterpreterDelegateTest.testDelegate
> 20:08:53  [       OK ] InterpreterDelegateTest.testDelegate
> 20:08:53  [ RUN      ] InterpreterDelegateTest.testDestructionOrder
> 20:08:53  [  FAILED  ] InterpreterDelegateTest.testDestructionOrder
> 20:08:53  [ RUN      ] InterpreterDelegateTest.testFail
> 20:08:53  Traceback (most recent call last):
> 20:08:53    File "_ctypes/callbacks.c", line 234, in 'calling callback function'
> 20:08:53    File "/.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/interpreter.py", line 111, in report
> 20:08:53      self.message += x
> 20:08:53  TypeError: must be str, not bytes
> 20:08:53  [  FAILED  ] InterpreterDelegateTest.testFail
> 20:08:53  [ RUN      ] InterpreterDelegateTest.testMultipleInterpreters
> 20:08:53  [       OK ] InterpreterDelegateTest.testMultipleInterpreters
> 20:08:53  [ RUN      ] InterpreterDelegateTest.testOptions
> 20:08:53  [       OK ] InterpreterDelegateTest.testOptions
> 20:08:53  [ RUN      ] InterpreterDelegateTest.test_session
> 20:08:53  W0821 19:08:46.430220 139904211531584 deprecation.py:323] From /usr/lib/python3.6/contextlib.py:60: TensorFlowTestCase.test_session (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
> 20:08:53  Instructions for updating:
> 20:08:53  Use `self.session()` or `self.cached_session()` instead.
> 20:08:53  [       OK ] InterpreterDelegateTest.test_session
> 20:08:53  [ RUN      ] InterpreterTensorAccessorTest.testBase
> 20:08:53  [       OK ] InterpreterTensorAccessorTest.testBase
> 20:08:53  [ RUN      ] InterpreterTensorAccessorTest.testBaseProtectsFunctions
> 20:08:53  [       OK ] InterpreterTensorAccessorTest.testBaseProtectsFunctions
> 20:08:53  [ RUN      ] InterpreterTensorAccessorTest.testGetTensorAccessor
> 20:08:53  [       OK ] InterpreterTensorAccessorTest.testGetTensorAccessor
> 20:08:53  [ RUN      ] InterpreterTensorAccessorTest.testTensorAccessor
> 20:08:53  [       OK ] InterpreterTensorAccessorTest.testTensorAccessor
> 20:08:53  [ RUN      ] InterpreterTensorAccessorTest.test_session
> 20:08:53  [       OK ] InterpreterTensorAccessorTest.test_session
> 20:08:53  [ RUN      ] InterpreterTest.testFloat
> 20:08:53  [       OK ] InterpreterTest.testFloat
> 20:08:53  [ RUN      ] InterpreterTest.testString
> 20:08:53  [       OK ] InterpreterTest.testString
> 20:08:53  [ RUN      ] InterpreterTest.testUint8
> 20:08:53  [       OK ] InterpreterTest.testUint8
> 20:08:53  [ RUN      ] InterpreterTest.test_session
> 20:08:53  [  SKIPPED ] InterpreterTest.test_session
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.testInvalidIndex
> 20:08:53  [       OK ] InterpreterTestErrorPropagation.testInvalidIndex
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.testInvalidModelContent
> 20:08:53  [       OK ] InterpreterTestErrorPropagation.testInvalidModelContent
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.testInvalidModelFile
> 20:08:53  [       OK ] InterpreterTestErrorPropagation.testInvalidModelFile
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.testInvalidModelFileContent
> 20:08:53  [       OK ] InterpreterTestErrorPropagation.testInvalidModelFileContent
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.testInvokeBeforeReady
> 20:08:53  [       OK ] InterpreterTestErrorPropagation.testInvokeBeforeReady
> 20:08:53  [ RUN      ] InterpreterTestErrorPropagation.test_session
> 20:08:53  [  SKIPPED ] InterpreterTestErrorPropagation.test_session
> 20:08:53  ======================================================================
> 20:08:53  FAIL: testDestructionOrder (__main__.InterpreterDelegateTest)
> 20:08:53  testDestructionOrder (__main__.InterpreterDelegateTest)
> 20:08:53  Make sure internal _interpreter object is destroyed before delegate.
> 20:08:53  ----------------------------------------------------------------------
> 20:08:53  Traceback (most recent call last):
> 20:08:53    File "/.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/interpreter_test.py", line 330, in testDestructionOrder
> 20:08:53      self.assertEqual(destructions, ['interpreter', 'test_delegate'])
> 20:08:53  AssertionError: Lists differ: ['interpreter', b'test_delegate'] != ['interpreter', 'test_delegate']
> 20:08:53  
> 20:08:53  First differing element 1:
> 20:08:53  b'test_delegate'
> 20:08:53  'test_delegate'
> 20:08:53  
> 20:08:53  - ['interpreter', b'test_delegate']
> 20:08:53  ?                 -
> 20:08:53  
> 20:08:53  + ['interpreter', 'test_delegate']
> 20:08:53  
> 20:08:53  ======================================================================
> 20:08:53  FAIL: testFail (__main__.InterpreterDelegateTest)
> 20:08:53  testFail (__main__.InterpreterDelegateTest)
> 20:08:53  ----------------------------------------------------------------------
> 20:08:53  Traceback (most recent call last):
> 20:08:53    File "/.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/interpreter.py", line 165, in load_delegate
> 20:08:53      delegate = Delegate(library, options)
> 20:08:53  ValueError
> 20:08:53  
> 20:08:53  During handling of the above exception, another exception occurred:
> 20:08:53  
> 20:08:53  ValueError: Failed to load delegate from /.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/testdata/test_delegate.so
> 20:08:53  
> 20:08:53  
> 20:08:53  During handling of the above exception, another exception occurred:
> 20:08:53  
> 20:08:53  Traceback (most recent call last):
> 20:08:53    File "/.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/interpreter_test.py", line 369, in testFail
> 20:08:53      self._delegate_file, options={'fail': 'fail'})
> 20:08:53  AssertionError: "Failed to load delegate from .*
> 20:08:53  Fail argument sent." does not match "Failed to load delegate from /.../execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/python/interpreter_test.runfiles/org_tensorflow/tensorflow/lite/python/testdata/test_delegate.so
> 20:08:53  "
> 20:08:53  
> 20:08:53  ----------------------------------------------------------------------
> 20:08:53  Ran 21 tests in 0.037s
> 20:08:53  
> 20:08:53  FAILED (failures=2, skipped=2)
> 20:08:53  ================================================================================
> 20:09:25  INFO: Elapsed time: 3377.988s, Critical Path: 296.39s
> 20:09:25  INFO: 4472 processes: 4472 local.
> 20:09:25  INFO: Build completed, 1 test FAILED, 4639 total actions
> 20:09:25  //tensorflow/lite/python:interpreter_test                                FAILED in 6.9s
> 20:09:25      FAILED  __main__.InterpreterDelegateTest.testDestructionOrder (0.0s)
> 20:09:25      FAILED  __main__.InterpreterDelegateTest.testFail (0.0s)
> 20:09:25  Test cases: finished with 130 passing and 2 failing out of 132 test cases
> 20:09:25  
> 20:09:25  Executed 9 out of 9 tests: 8 tests pass and 1 fails locally.
> 20:09:25  There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
> 20:09:25  INFO: Build completed, 1 test FAILED, 4639 total actions
>

### CODE REVIEW

**Review by aselle — COMMENTED** [review]
Thanks for the contribution.  Could you please make the suggested changes?

File: tensorflow/lite/python/interpreter.py [hunk 1]
```diff
@@ -108,7 +108,11 @@ def __init__(self):
         self.message = ''
 
       def report(self, x):
-        self.message += x
+        if (type(x) is not bytes):
+          self.message += x
+        else:
+          # To support both: PY2 and PY3, we don't use the function decode, but do this way
+          self.message.join(chr(l) for l in x)
```
[the remark this answers is not in the document]
Comment (akarmi) [reply tensorflow/lite/python/interpreter.py]: Thanks, it should be fixed now

File: tensorflow/lite/python/interpreter_test.py [hunk 2]
```diff
@@ -342,8 +347,8 @@ def testOptions(self):
 
     delegate_b = interpreter_wrapper.load_delegate(
         self._delegate_file, options={
-            'unused': False,
-            'options_counter': 2
+          'unused': False,
```
[the remark this answers is not in the document]
Comment (akarmi) [reply tensorflow/lite/python/interpreter_test.py]: Thanks, we reverted this change.

**Comment by gbaned** [comment]
@akarmi Could you please resolve the conflicts? Thanks!

**Comment by akarmi** [comment]
It appears that the main issue has recently been resolved by [this commit](https://github.com/tensorflow/tensorflow/commit/b608d1180567fd303107a9381bbdf7dd7555d9f4). I resolved the conflict. Now, this PR only fixes an issue with the invalid pointer in tflite_plugin_destroy_delegate() and exception message filtering in one of the test assert statements.

**Comment by akarmi** [comment]
@aselle, can you please review this PR? It's been outstanding for quite some time. I believe it should be merged to ensure the complete resolution of the described issues. Thanks.

**Review by jdduke — APPROVED** [review]

**Pull request merged**